core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java

```diff
@@ -3090,6 +3102,13 @@ private static class ItemImplementor extends AbstractRexCallImplementor {
         final RexCall call, final List<Expression> argValueList) {
       final MethodImplementor implementor =
           getImplementor(call.getOperands().get(0).getType().getSqlTypeName());
+      // If the structure is an array, two additional arguments are added to the argValueList
+      // to denote what the offset is and what the behavior should be if the index is out of boudns.
+      if (implementor.method.getName().equals("arrayItemOptional")) {
```

Comment:
Hard-coded method names are too fragile.

Especially strange given that getImplementor returns BuiltInMethod.ARRAY_ITEM and there's no BuiltInMethod.ARRAY_ITEM_OPTIONAL.

Reading this code, I first thought it was because of the 'optional'. But actually it's because it's an array.

So, figure out a way to make this code more robust and intuitive. It's not sufficient that it works.

Reply:
I added a commit that tries to tidy this up by checking the SqlTypeName of the collection argument instead of checking the implementor method. Let me know what you think.